chore: update modules

stevenh opened the pull request:
Update modules to latest versions, addressing dependabot alerts.

stevenh committed:
chore: update modules

Update tools and modules to latest versions, addressing dependabot
alerts.

stevenh committed:
chore: env debug

Add go version and env debug for issue raising.

stevenh committed:
chore: bug test

Add a standalone test for the os.ReadDir issue.

stevenh committed:
chore: update module dependencies

Update all module dependencies and go to the latest versions.

stevenh committed:
chore: update github actions to latest versions

Update GitHub Actions workflows to use the latest versions of actions
to address issue with golang 1.24 compatibility.

stevenh commented:
Waiting on a release with https://github.com/golang/go/issues/75157 in it, to fix the tests.

dmitshur commented on pkg/gopro/osfs_test.go:
```diff
@@ -34,7 +47,8 @@ func TestOSFS(t *testing.T) {
 	require.ErrorIs(t, err, syscall.ENOTDIR)
-	require.Equal(t, []fs.DirEntry{}, dirs)
+	var expectedDirs []fs.DirEntry
+	require.Equal(t, expectedDirs, dirs)
```
Consider this modified test case:

```diff
-require.Equal(t, []fs.DirEntry{}, dirs)
+require.Equal(t, 0, len(dirs))
```

It still checks that the number of directories returned is as expected, zero, but it's not sensitive to whether the slice happens to be empty and non-nil, or nil.

The documentation at https://pkg.go.dev/io/fs#ReadDirFS.ReadDir states only that ReadDir "returns a list of directory entries". The documentation doesn't specify that a slice is nil or non-nil.

Since you're implementing your own interface, you could choose to add logic to detect when the slice has 0 elements and consistently make it a nil slice, but I suspect that behavior isn't going to be helpful to most callers that would at most `range` over the slice, and ranging over a nil slice behaves no different than ranging over an empty slice. Also see the Go style suggestion at https://go.dev/s/style#declaring-empty-slices:

> When designing interfaces, avoid making a distinction between a nil slice and a non-nil, zero-length slice

dmitshur commented on pkg/gopro/osfs_test.go:
```diff
@@ -10,6 +10,19 @@ import (
+	dirs, err := os.ReadDir(name)
+	require.ErrorIs(t, err, syscall.ENOTDIR)
+	var expectedDirs []fs.DirEntry
+	require.Equal(t, expectedDirs, dirs)
```
I realize this test is probably added in the context of issue golang/go#75157. Also noting here that https://pkg.go.dev/os#ReadDir doesn't specify whether the returned slice is nil or not. Its documented behavior only promises that "If an error occurs reading the directory, ReadDir returns the entries it was able to read before the error, along with the error."

dmitshur reviewed (commented):
I left a few suggestions on how the test cases can be updated to not run into issue golang/go#75157. It should work without needing to skip those tests.

stevenh committed:
chore: switch to require.Empty for slice checks

Switch to using require.Empty for slice checks to avoid failure due to
inconsistent behaviour of ReadDir between platforms
